File: bench.py
"""bench.py - speed benchmark: resprpy vs respR (same data from CSV)."""
import time

import numpy as np
import resprpy as p

BASE = "tests/reference/"


def bench(label, fn, repeat=2):
    best = float("inf")
    for _ in range(repeat):
        t0 = time.perf_counter()
        fn()
        best = min(best, time.perf_counter() - t0)
    print(f"{label} {best:.4f}")


u = np.genfromtxt(BASE + "urchins.csv", delimiter=",", skip_header=1)
d2k = np.genfromtxt("reference/bench_2k.csv", delimiter=",", skip_header=1)
d10k = np.genfromtxt("reference/bench_10k.csv", delimiter=",", skip_header=1)

bench("calc_rate_271", lambda: p.calc_rate(u[:, :2], from_=0, to=45, by="time"))
bench("calc_rate_2k", lambda: p.calc_rate(d2k, from_=0, to=1800, by="time"))
bench("calc_rate_10k", lambda: p.calc_rate(d10k, from_=0, to=9000, by="time"))
bench("auto_rate_271", lambda: p.auto_rate(u[:, :2]))
bench("auto_rate_2k", lambda: p.auto_rate(d2k))
bench("auto_rate_10k", lambda: p.auto_rate(d10k))

File: reference/bench.R
# bench.R - speed benchmark: respR vs pyrespr (same data, saved to CSV)
suppressPackageStartupMessages(library(respR))
u <- read.csv("../tests/reference/urchins.csv")
mk <- function(n, seed) {
  set.seed(seed)
  t <- 0:(n - 1)
  o <- 8 - 0.001 * t + rnorm(n, 0, 0.01)
  data.frame(time = t, oxy = o)
}
d2k <- mk(2000, 1)
d10k <- mk(10000, 2)
write.csv(d2k, "bench_2k.csv", row.names = FALSE)
write.csv(d10k, "bench_10k.csv", row.names = FALSE)

bench <- function(label, expr) {
  tm <- system.time(expr)[3]
  cat(sprintf("%s %.4f\n", label, tm))
}

bench("calc_rate_271", calc_rate(u[, 1:2], from = 0, to = 45, by = "time",
                                 plot = FALSE))
bench("calc_rate_2k", calc_rate(d2k, from = 0, to = 1800, by = "time",
                                plot = FALSE))
bench("calc_rate_10k", calc_rate(d10k, from = 0, to = 9000, by = "time",
                                 plot = FALSE))
bench("auto_rate_271", auto_rate(u[, 1:2], plot = FALSE))
bench("auto_rate_2k", auto_rate(d2k, plot = FALSE))
bench("auto_rate_10k", auto_rate(d10k, plot = FALSE))
